src/tls/test/main.cpp

```diff
@@ -0,0 +1,376 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the Apache 2.0 License.
+#include "crypto/certs.h"
+#include "crypto/key_pair.h"
+#include "crypto/verifier.h"
+#define DOCTEST_CONFIG_IMPLEMENT_WITH_MAIN
+#include "ds/buffer.h"
+#include "ds/logger.h"
+#include "tls/client.h"
+#include "tls/msg_types.h"
+#include "tls/server.h"
+
+#include <chrono>
+#include <doctest/doctest.h>
+#include <iostream>
+#include <memory>
+#include <string>
+#include <sys/socket.h>
+#include <thread>
+
+using namespace std;
+using namespace crypto;
+using namespace tls;
+
+logger::ConsoleLogger test_log;
+
+/// Server uses one descriptor while client uses the other.
+/// Use the send/recv template wrappers below as callbacks.
+class TestPipe
+{
+  int pfd[2];
+
+public:
+  static const int SERVER = 0;
+  static const int CLIENT = 1;
+
+  TestPipe(bool dgram)
+  {
+    int sock_type = 0;
+    if (dgram)
+      sock_type = SOCK_DGRAM;
+    else
+      sock_type = SOCK_STREAM;
+    if (socketpair(PF_LOCAL, sock_type, 0, pfd) == -1)
+    {
+      throw runtime_error(
+        "Failed to create socketpair: " + string(strerror(errno)));
+    }
+  }
+  ~TestPipe()
+  {
+    close(pfd[0]);
+    close(pfd[1]);
+  }
+
+  size_t send(int id, const uint8_t* buf, size_t len)
+  {
+    return write(pfd[id], buf, len);
+  }
+
+  size_t recv(int id, uint8_t* buf, size_t len)
+  {
+    return read(pfd[id], buf, len);
+  }
+};
+
+/// Callback wrapper around TestPipe->send().
+template <int end>
+int send(void* ctx, const uint8_t* buf, size_t len)
+{
+  auto pipe = reinterpret_cast<TestPipe*>(ctx);
+  return pipe->send(end, buf, len);
+}
+
+/// Callback wrapper around TestPipe->recv().
+template <int end>
+int recv(void* ctx, uint8_t* buf, size_t len)
+{
+  auto pipe = reinterpret_cast<TestPipe*>(ctx);
+  return pipe->recv(end, buf, len);
+}
+
+/// mbedtls debug call back.
+static void dbg_callback(
+  void*, int, const char* file, int line, const char* str)
+{
+  test_log.write(string(file) + ":" + to_string(line) + " " + str);
```

Comment:
We use fmtlib (https://github.com/fmtlib/fmt) elsewhere in CCF, it's quite good. This would be `fmt::format("{}:{} {}", file, line, str)` for example.

Reply:
Right, I was just writing quick logs to help debug, I'll change to use `fmt` and rebase.